tests/functional/test_linear_tracer.py

```diff
@@ -0,0 +1,262 @@
+
+    # NOTE: Due to the hyperbolic nature, the error in the tracer fraction should
+    # converge to zero linearly, which is checked in a separate test. Here we check only
+    # that the error is not getting worse with ongoing code development.
+    expected_error_z = [
+        0.1128851432293962,
+        0.1344878688028854,
+        0.1489878242519587,
+        0.15299862798949995,
+    ]
+    expected_error_diffused_z = [
+        0.020278908152713,
+        0.018535428299520068,
+        0.01703231859562048,
+        0.016602357208773923,
+    ]
+    # Allow 10% margin for machine reasons.
+    assert (
+        sol_data.error_diffused_z_tracer <= 1.1 * expected_error_diffused_z[time_index]
+    )
+    assert sol_data.error_z_tracer <= 1.1 * expected_error_z[time_index]
+
+
+# Expected to fail because Upwinding and backward euler are diffusive
+@pytest.mark.xfail
+@pytest.mark.parametrize(
+    "results",
+    [
+        (SimplePipe2D.pipe_length / 40, TracerFlowSetup_1p),
+        (SimplePipe2D.pipe_length / 40, TracerFlowSetup_1p_ff),
+    ],
+    indirect=["results"],
+)
+@pytest.mark.parametrize("time_index", [0, 1, 2, 3])
+def test_linear_tracer_1p_is_diffusive(
+    time_index: int, results: list[LinearTracerSaveData]
+) -> None:
+    """Compares the tracer fraction with the exact solution with a steep front.
+
+    This test is expected to fail since the numerical algorithm used is diffusive.
+
+    """
+    sol_data = results[time_index]
+    np.testing.assert_allclose(sol_data.error_z_tracer, 0.0, atol=1e-7, rtol=0.0)
+
+
+@pytest.mark.skipped  # reason: slow
+@pytest.mark.parametrize("model_class", [TracerFlowSetup_1p, TracerFlowSetup_1p_ff])
+def test_linear_tracer_1p_ooc(
+    model_class: type[TracerFlowSetup_1p] | type[TracerFlowSetup_1p_ff],
+) -> None:
+    """Tests the order of convergence for the tracer fraction, which is expected to be
+    slightly super-linear in the L1 norm, and roughly linear in the L2 norm."""
+    max_iterations = 80
+    newton_tol = 1e-6
+    newton_tol_increment = newton_tol
+
+    # Breakthrough time is roughly 100 seconds in this setup. We take the half time
+    # where the front is roughly in the middle and both sides of the modified solution
+    # which includes diffusiones are shown in the solution
+    time_manager = pp.TimeManager([0, 50], 10, True)
+
+    params = {
+        "material_constants": {
+            "solid": pp.SolidConstants(
+                porosity=1, permeability=1, residual_aperture=1.0
+            ),
+        },
+        "time_manager": time_manager,
+        "max_iterations": max_iterations,
+        "nl_convergence_tol": newton_tol_increment,
+        "nl_convergence_tol_res": newton_tol,
+        "times_to_export": [],
+        "grid_type": "cartesian",
+        "meshing_arguments": {"cell_size": 1.0},
+    }
+
+    conv_analysis = ConvergenceAnalysis(
+        model_class=model_class,
+        model_params=deepcopy(params),
+        levels=5,
```

Comment:
Is five levels really necessary? From my tests, it seems the region of convergence is reached after 3, or at most 4, levels.

Reply:
Switched to 3 levels of refinement, added test with spatial refinement only.